grblas/base.py

```diff
@@ -125,13 +233,128 @@ def new(self, *, dtype=None, mask=None):
+        self.parent[self.resolved_indexes] = obj
+
+    @property
+    def value(self):
+        if isinstance(self.parent, Updater):
+            raise TypeError('Cannot extract from an Updater')
+        if not self.resolved_indexes.is_single_element():
+            raise AttributeError("Only Scalars have `.value` attribute")
+        val, _ = self.parent._extract_element(self.resolved_indexes)
+        return val
+
+    def new(self, *, dtype=None, mask=None):
+        """
+        Force extraction of the indexes into a new object
+        dtype and mask are the only controllable parameters.
+        """
+        if isinstance(self.parent, Updater):
+            raise TypeError('Cannot extract from an Updater')
+        if self.resolved_indexes.is_single_element():
+            if mask is not None:
+                raise TypeError('mask is not allowed for single element extraction')
+            val, cur_dtype = self.parent._extract_element(self.resolved_indexes)
+            if dtype is None:
+                dtype = cur_dtype
+            from .scalar import Scalar
+            return Scalar.new_from_value(val, dtype=dtype)
+        else:
+            delayed_extractor = self.parent._prep_for_extract(self.resolved_indexes)
+            return delayed_extractor.new(dtype=dtype, mask=mask)
+
+    def _extract_delayed(self):
+        """Return a GbDelayed object, treating this as an extract call"""
+        if isinstance(self.parent, Updater):
+            raise TypeError('Cannot extract from an Updater')
+        if self.resolved_indexes.is_single_element():
+            raise TypeError('extract and update is not allowed for single element')
+        return self.parent._prep_for_extract(self.resolved_indexes)
+
+
+class IndexerResolver:
+    def __init__(self, obj, indices):
+        if obj.is_scalar:
+            raise TypeError("Cannot index into Scalars")
+        self.obj = obj
+        self.indices = self.parse_indices(indices, obj.shape)
+
+    def is_single_element(self):
+        for idx, size in self.indices:
+            if size is not None:
+                return False
+        return True
+
+    def parse_indices(self, indices, shape):
+        """
+        Returns
+            [(rows, rowsize), (cols, colsize)] for Matrix
+            [(idx, idx_size)] for Vector
+
+        Within each tuple, if the index is of type int, the size will be None
+        """
+        if len(shape) == 1:
+            if type(indices) == tuple:
+                raise TypeError(f'Index for {self.obj.__class__.__name__} cannot be a tuple')
+            # Convert to tuple for consistent processing
+            indices = (indices,)
+        elif len(shape) == 2:
+            if type(indices) != tuple or len(indices) != 2:
+                raise TypeError(f'Index for {self.obj.__class__.__name__} must be a 2-tuple')
+
+        out = []
+        for i, idx in enumerate(indices):
+            typ = type(idx)
+            if type == tuple:
```

Comment:
Did you mean `typ` instead of `type`?  Also, `is`.

Reply:
Good catch